File: com/jedi/ObiWan.java
package com.jedi;

import org.jruby.Ruby;
import org.jruby.RubyObject;
import org.jruby.javasupport.util.RuntimeHelpers;
import org.jruby.runtime.builtin.IRubyObject;
import org.jruby.javasupport.JavaUtil;
import org.jruby.RubyClass;


public class ObiWan extends RubyObject  {
    private static final Ruby __ruby__ = Ruby.getGlobalRuntime();
    private static final RubyClass __metaclass__;

    static {
        String source = new StringBuilder("require 'java'\n" +
            "\n" +
            "java_package 'com.jedi'\n" +
            "\n" +
            "class Jedi\n" +
            "\n" +
            "  java_signature 'Jedi(String name, String lightsaber_color)'\n" +
            "  def initialize(name, lightsaber_color)\n" +
            "    @name= name\n" +
            "    @lightsaber_color= lightsaber_color\n" +
            "  end\n" +
            "\n" +
            "  # Note: using explicit readers for now to add java_signature\n" +
            "\n" +
            "  java_signature 'String name()'\n" +
            "  def name\n" +
            "    @name\n" +
            "  end\n" +
            "\n" +
            "  java_signature 'String lightsaber_color()'\n" +
            "  def lightsaber_color\n" +
            "    @lightsaber_color\n" +
            "  end\n" +
            "\n" +
            "end\n" +
            "\n" +
            "class ObiWan < Jedi\n" +
            "\n" +
            "  java_signature 'ObiWan()'\n" +
            "  def initialize\n" +
            "    super('Obi Wan Kenobi', 'blue')\n" +
            "  end\n" +
            "\n" +
            "end\n" +
            "\n" +
            "class LukeSkyWalker < Jedi\n" +
            "\n" +
            "  java_signature 'LukeSkywalker()'\n" +
            "  def initialize\n" +
            "    super('Luke Skywalker', 'green')\n" +
            "  end\n" +
            "\n" +
            "end\n" +
            "\n" +
            "#obi = ObiWan.new\n" +
            "#luke = LukeSkyWalker.new\n" +
            "#\n" +
            "#\n" +
            "#jedis = java.util.ArrayList.new\n" +
            "#jedis << obi\n" +
            "#jedis << luke\n" +
            "#\n" +
            "#jedis.each do |j|\n" +
            "#  puts \"-------Jedi-----------------------------\"\n" +
            "#  puts \"name: #{j.name}\"\n" +
            "#  puts \"lightsaber_color: #{j.lightsaber_color}\"\n" +
            "#  puts \"----------------------------------------\"\n" +
            "#end\n" +
            "\n" +
            "\n" +
            "\n" +
            "\n" +
            "").toString();
        __ruby__.executeScript(source, "obi_wan.rb");
        RubyClass metaclass = __ruby__.getClass("ObiWan");
        metaclass.setRubyStaticAllocator(ObiWan.class);
        if (metaclass == null) throw new NoClassDefFoundError("Could not load Ruby class: ObiWan");
        __metaclass__ = metaclass;
    }

    /**
     * Standard Ruby object constructor, for construction-from-Ruby purposes.
     * Generally not for user consumption.
     *
     * @param ruby The JRuby instance this object will belong to
     * @param metaclass The RubyClass representing the Ruby class of this object
     */
    private ObiWan(Ruby ruby, RubyClass metaclass) {
        super(ruby, metaclass);
    }

    /**
     * A static method used by JRuby for allocating instances of this object
     * from Ruby. Generally not for user comsumption.
     *
     * @param ruby The JRuby instance this object will belong to
     * @param metaclass The RubyClass representing the Ruby class of this object
     */
    public static IRubyObject __allocate__(Ruby ruby, RubyClass metaClass) {
        return new ObiWan(ruby, metaClass);
    }

    
    public  ObiWan() {
        this(__ruby__, __metaclass__);

        RuntimeHelpers.invoke(__ruby__.getCurrentContext(), this, "initialize");

    }

}

File: obi_wan.rb
include Java
require 'Destroyer.jar'

#require 'java'
#
#java_package 'com.jedi'

#class Jedi
#
#  java_signature 'Jedi(String name, String lightsaber_color)'
#  def initialize(name, lightsaber_color)
#    @name= name
#    @lightsaber_color= lightsaber_color
#  end
#
#  # Note: using explicit readers for now to add java_signature
#
#  java_signature 'String name()'
#  def name
#    @name
#  end
#
#  java_signature 'String lightsaber_color()'
#  def lightsaber_color
#    @lightsaber_color
#  end
#
#end
#
#class ObiWan < Jedi
#
#  java_signature 'ObiWan()'
#  def initialize
#    super('Obi Wan Kenobi', 'blue')
#  end
#
#end
#
#class LukeSkyWalker < Jedi
#
#  java_signature 'LukeSkywalker()'
#  def initialize
#    super('Luke Skywalker', 'green')
#  end
#
#end

#obi = ObiWan.new
#luke = LukeSkyWalker.new
#
#
#jedis = java.util.ArrayList.new
#jedis << obi
#jedis << luke
#
#jedis.each do |j|
#  puts "-------Jedi-----------------------------"
#  puts "name: #{j.name}"
#  puts "lightsaber_color: #{j.lightsaber_color}"
#  puts "----------------------------------------"
#end

class ObiWan
  include Java::ComJedi::Jedi

  def initialize(name, lightsaber_color)
    @name= name
    @lightsaber_color= lightsaber_color
  end

  java_signature 'String name()'
  def name
    @name + " is NOT the best"
  end

  java_signature 'String lightsaber_color()'
  def lightsaber_color
    @lightsaber_color
  end

  java_signature 'String showBytes(byte [] data)'
  def show_bytes(data)
    puts String.from_java_bytes(data)
  end

end

java_import 'com.jedi.Destroyer'

obi_wan = ObiWan.new 'ObiWan', 'blue'
destroyer = Destroyer.new obi_wan

puts destroyer.getJedi().lightsaber_color
puts destroyer.showJedi
puts

puts "Java bytes:\n\n"

bytes = destroyer.getBytes()
puts bytes.inspect
bytes.each_with_index do |b,i|
  puts "#{i} => #{b} -> %#X" % b
end

puts

puts "Ruby bytes:\n\n"

ruby_bytes = bytes.to_a
puts ruby_bytes.inspect

ruby_bytes.to_a.each_with_index do |b,i|
  puts "#{i} => #{b} -> %#X" % b
end

puts
puts "Ruby bytes given to Java to print\n\n"
destroyer.showBytes(bytes)

puts "Serialization:"
puts

packed = ruby_bytes.pack("C*").unpack("H*")
puts "Packed: " + packed.first

unpacked = packed.pack("H*").unpack("c*")
puts "Unpacked: " + unpacked.inspect
puts

puts "Unpacked bytes given to java to print"

destroyer.showBytes(unpacked)


